### PR TITLE
Add an example

### PR DESCRIPTION
**Pull request by serg06**
I had a really tough time figuring out how to do simple multi-threaded pub-sub with cppzmq; I finally figured it out so I'd like to add it.

### CODE REVIEW

File: examples/pubsub_multithread_inproc.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+int main()
+{
+	/*
+	 * No I/O threads are involved in passing messages using the inproc transport.
+	 * Therefore, if you are using a ØMQ context for in-process messaging only you
+	 * can initialise the context with zero I/O threads.
+	 *
+	 * Source: http://api.zeromq.org/2-1:zmq-inproc
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:83]: Maybe link to a more recent version http://api.zeromq.org/4-3:_start

File: examples/pubsub_multithread_inproc.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	 *     Thread2: [A] Message in A envelope
+	 *     Thread2: [B] Message in B envelope
+	 *     Thread3: [A] Message in A envelope
+	 *     Thread3: [B] Message in B envelope
+	 *     Thread3: [C] Message in C envelope
+	 */
+}
+
+#ifdef _WIN32
+int CALLBACK WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:110]: is this really required?

File: examples/pubsub_multithread_inproc.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		zmq::message_t address;
+		zmq::recv_result_t result = subscriber.recv(address);
+
+		//  Read message contents
+		zmq::message_t contents;
+		result = subscriber.recv(contents);
+
+		std::stringstream out;
+		out << "Thread3: [" << address.to_string_view() << "] " << contents.to_string_view() << std::endl;
+		OutputDebugString(out.str().c_str());
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:72]: Use `std::cout` directly instead.

File: examples/pubsub_multithread_inproc.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	//  Prepare our context and subscriber
+	zmq::socket_t subscriber(*ctx, zmq::socket_type::sub);
+	subscriber.connect("inproc://#1");
+
+	//  Thread3 opens ALL envelopes
+	subscriber.setsockopt(ZMQ_SUBSCRIBE, "", 0);
+
+	while (1) {
+		//  Read envelope address
+		zmq::message_t address;
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:63]: A better name would be `topic` in my opinion

File: examples/pubsub_multithread_inproc.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		zmq::recv_result_t result = subscriber.recv(address);
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:64]: I think there should be some error checking, at least an `assert`. You could also use `zmq::recv_multipart` to simplify the message handling.

File: examples/pubsub_multithread_inproc.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+void SubscriberThread2(zmq::context_t* ctx)
+{
+	//  Prepare our context and subscriber
+	zmq::socket_t subscriber(*ctx, zmq::socket_type::sub);
+	subscriber.connect("inproc://#1");
+
+	//  Thread3 opens ALL envelopes
+	subscriber.setsockopt(ZMQ_SUBSCRIBE, "", 0);
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:59]: The subscribes can use the type safe way `subscriber.set(zmq::sockopt::subscribe, "");`

File: examples/pubsub_multithread_inproc.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+void SubscriberThread1(zmq::context_t* ctx)
+{
+	//  Prepare subscriber
+	zmq::socket_t subscriber(*ctx, zmq::socket_type::sub);
+	subscriber.connect("inproc://#1");
+
+	//  Thread2 opens "A" and "B" envelopes
+	subscriber.setsockopt(zmq::sockopt::subscribe, "A", 1);
+	subscriber.setsockopt(zmq::sockopt::subscribe, "B", 1);
```
Comment (gummif) [inline examples/pubsub_multithread_inproc.cpp:37]: I think that should be `subscriber.set(zmq::sockopt::subscribe, "B");`

Comment (serg06) [reply examples/pubsub_multithread_inproc.cpp]: Oh my bad, I misread the first time. Fixed.

**Comment by sigiesec** [comment]
Thanks for adding the examples, that's really helpful! Could you just reformat that using clang-format?

**Comment by sigiesec** [comment]
Thanks! Sorry for taking so long to merge this.

**Pull request merged**